torchdynamo/_eval_frame.c

```diff
@@ -310,6 +286,23 @@ inline static PyObject *eval_custom_code(PyThreadState *tstate,
   return result;
 }
 
+static PyObject *_custom_eval_frame_shim(PyThreadState *tstate, PyFrameObject *frame,
+                                    int throw_flag) {
+// Shims logic into one of three states. Can probably be refactored into a single func, later:
+  //  - None: disables TorchDynamo
+  //  - False: run-only mode (reuse existing compiles)
+  //  - Python callable(): enables TorchDynamo
+    PyObject *callback = eval_frame_callback_get();
```

Comment:
Now that we have this shim, perhaps we should change the signature of _custom_eval_frame to accept `callback` as an arg, this would let us avoid a thread-local-storage lookup.

Reply:
Sure, let me try that.

Reply:
Actually, I think I don't understand, but I probably am missing knowledge about a critical mechanism.

This is called from the eval_frame, so this shim gets set on via `tstate->interp->eval_frame = &custom_eval_frame_shim;`. How can we pass in a callback to this as an arg?

Reply:
See https://github.com/pytorch/torchdynamo/pull/239/files#r873054114